p2p/net/conn/dial_test.go

```diff
@@ -397,3 +399,221 @@ func TestFailedAccept(t *testing.T) {
+		}
+		defer con2.Close()
+
+		err = msmux.SelectProtoOrFail(SecioTag, con2)
+		if err != nil {
+			t.Error("msmux select failed: ", err)
+		}
+
+		_, err = con2.Write([]byte("test"))
+		if err != nil {
+			t.Error("con write failed: ", err)
+		}
+	}()
+
+	c, err := l1.Accept()
+	if err != nil {
+		t.Fatal("connections after a failed accept should still work: ", err)
+	}
+
+	c.Close()
+	<-done
+}
+
+func TestConcurrentAccept(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	p1 := tu.RandPeerNetParamsOrFatal(t)
+
+	l1, err := Listen(ctx, p1.Addr, p1.ID, p1.PrivKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	n := 300
+	delay := time.Millisecond * 20
+	if runtime.GOOS == "darwin" {
+		n = 100
+	}
+
+	p1.Addr = l1.Multiaddr() // Addr has been determined by kernel.
+
+	var wg sync.WaitGroup
+	for i := 0; i < n; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			con, err := net.Dial("tcp", l1.Addr().String())
+			if err != nil {
+				log.Error(err)
+				t.Error("first dial failed: ", err)
+				return
+			}
+			// hang this connection
+			defer con.Close()
+
+			time.Sleep(delay)
+			err = msmux.SelectProtoOrFail(SecioTag, con)
+			if err != nil {
+				t.Error(err)
+			}
+		}()
+	}
+
+	before := time.Now()
+	for i := 0; i < n; i++ {
+		c, err := l1.Accept()
+		if err != nil {
+			t.Fatal("connections after a failed accept should still work: ", err)
+		}
+
+		c.Close()
+	}
+
+	limit := delay * time.Duration(n)
+	took := time.Now().Sub(before)
+	if took > limit {
+		t.Fatal("took too long!")
+	}
+	log.Errorf("took: %s (less than %s)", took, limit)
+}
+
+func TestConnectionTimeouts(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	old := NegotiateReadTimeout
+	NegotiateReadTimeout = time.Second * 5
+	defer func() { NegotiateReadTimeout = old }()
+
+	p1 := tu.RandPeerNetParamsOrFatal(t)
+
+	l1, err := Listen(ctx, p1.Addr, p1.ID, p1.PrivKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	n := 100
+	if runtime.GOOS == "darwin" {
+		n = 50
+	}
+
+	p1.Addr = l1.Multiaddr() // Addr has been determined by kernel.
+
+	var wg sync.WaitGroup
+	for i := 0; i < n; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			con, err := net.Dial("tcp", l1.Addr().String())
+			if err != nil {
+				log.Error(err)
+				t.Error("first dial failed: ", err)
+				return
+			}
+			defer con.Close()
+
+			// hang this connection until timeout
+			io.ReadFull(con, make([]byte, 1000))
+		}()
+	}
+
+	// wait to make sure the hanging dials have started
+	time.Sleep(time.Millisecond * 50)
+
+	good_n := 20
+	for i := 0; i < good_n; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			con, err := net.Dial("tcp", l1.Addr().String())
+			if err != nil {
+				log.Error(err)
+				t.Error("first dial failed: ", err)
+				return
+			}
+			defer con.Close()
+
+			// dial these ones through
+			err = msmux.SelectProtoOrFail(SecioTag, con)
+			if err != nil {
+				t.Error(err)
+			}
```

Comment:
(like, Write `good` from here, and check that comes through on the N conns accepted on the listener side)